feat: health check

folke opened the pull request:
Health check that tests some basic things for a user to troubleshoot fzf-lua.

~~There's a bug in Neovim that prevents the healthcheck to work, when its in `lua/fzf-lua/health.lua`.~~
~~For now I've put it in `lua/fzf/health.lua`, but this is of course **not** ideal.~~
~~The issue can be fixed upstream, but then non-nightly Neovim installs would fail running the health checks...~~

You can test it with `:checkhealth fzf_lua`.

This is how it looks for me:

![image](https://github.com/ibhagwan/fzf-lua/assets/292349/30c9a0f8-5f5a-47eb-b988-ba9b5ad0d5cd)

ibhagwan commented:
Looks great @folke.

A few observations:
- Given we decided we're accepting that some users reuse existing `$FZF_DEFAULT_OPTS` should we warn about this?
- Fzf-lua doesn't require `fd`, `sk`, `bat` or `delta`, it can work with the default `grep`  on Linux/Mac and requires `rg` only on windows and can list files with the default `find` or `dir.exe` on Windows.
- It seems that health only supports `OK` or `WARNING`, in the above case does it make sense to warn about missing `delta`  given that it's totally optional?

folke commented:
Added the additional checks per OS and removed the warning:

![image](https://github.com/ibhagwan/fzf-lua/assets/292349/05343820-183b-485a-afe3-ddff27a24a2c)

folke commented:
Requirements are now correctly checked for each OS

folke commented:
I was able to work-around that Neovim bug with health.
Instead of adding `lua/fzf/health.lua`, I've moved the heath check to `autoload/health/fzf_lua.vim`.

folke commented:
Added an extra check for stdpath.run.
See also https://github.com/ibhagwan/fzf-lua/issues/1243#issuecomment-2168891260

ibhagwan commented:
> I was able to work-around that Neovim bug with health.
Instead of adding lua/fzf/health.lua, I've moved the heath check to autoload/health/fzf_lua.vim.
Which then calls lua/fzf-lua/_health.lua. Note that the file can't be named health.lua, or it will trigger errors.
> You can use :checkhealth fzf_lua to just check for fzf-lua

That’s great @folke, tysm!

> Added an extra check for stdpath.run.

Good idea, I’ll test this on both Mac and Windows.

ibhagwan commented on lua/fzf-lua/_health.lua:
```diff
@@ -0,0 +1,115 @@
+  local required = {
+    { "fzf", "sk" },
+    is_win and { "rg" } or { "rg", "grep" },
+    is_win and { "fd", "find", "dir" } or { "fd", "find" },
```
I’ve seen Linux distributions that name `fd` as `fdfind`, hence:
https://github.com/ibhagwan/fzf-lua/blob/06d4fe9f9f86c40756b665b535b7d1dd6f25b639/lua/fzf-lua/providers/files.lua#L29-L30

folke replied:
I see you already added this :)

ibhagwan commented on lua/fzf-lua/_health.lua:
```diff
@@ -0,0 +1,115 @@
+  end
+
+  start("fzf-lua [optional:media]")
+  for _, tool in ipairs({ "chafa", "viu", "ueberzugpp" }) do
```
Do you think the media files check should be limited to Linux/Mac? I don’t think these tools can be compiled on Windows.

Unrelated to this PR: Now that I think if it, if we can auto detect a media previewer binary installed, might be a good idea  to set it automatically.

folke replied:
updated!

ibhagwan commented on lua/fzf-lua/_health.lua:
```diff
@@ -0,0 +1,115 @@
+  end
+
+  local run = vim.fn.stdpath("run")
+  if vim.fn.isdirectory(run) == 0 then
```
I’ve seen cases where the runtime was available but wasn’t writeable, access check here might be worth it too, I’m thinking creating a temp file with `os.tmpname` or `vim.fn.tempname` (runs dry, requires additional “touch”) and then deleting the temp file?

folke replied:
I added a check with `uv.fs_access` for `rxw`

ibhagwan commented:
Added some more ideas in the code, lmk what you think or if you’d like me to add the proposed changes.

folke commented:
You probably forgot to push your changes? I don't see anything here, no review or code updates.
But for sure, change it to your liking :)

ibhagwan commented:
> You probably forgot to push your changes? I don't see anything here, no review or code updates. But for sure, change it to your liking :)

quite frankly, I’ve never done a proper “review” I just wanted to add comments with line context and it opened a review, I usually do a manual rebase, modify, and change commits on the command line, never figured how to propagate these back to an open PR lol

folke commented:
When you add comments in a review, at the end you need to click top right, to submit the review, so that the comments show up

ibhagwan commented:
> When you add comments in a review, at the end you need to click top right, to submit the review, so that the comments show up

Gonna try it for fun :)

folke commented:
With the gh tool, you can actually make changes to a PR yourself.

First do `gh pr checkout 123`

Test, make changes and you can then do a simple git push to update the PR. Super useful. I do that all the time

ibhagwan commented:
> With the gh tool, you can actually make changes to a PR yourself.
> 
> First do `gh pr checkout 123`
> 
> Test, make changes and you can then do a simple git push to update the PR. Super useful. I do that all the time

Oh nice, I just stick to vanilla git with an alias to pull PRs locally but this is a good reason to use the tool
```gitconfig
  fpr   = "!f(){ git fetch origin pull/${@}/head:pr_${@}; }; f"
```

I was wondering why people never seem to see my code comments lol

ibhagwan commented:
Hi @folke, just saw your changes for the directory writeable test, I also added https://github.com/ibhagwan/fzf-lua/pull/1267/commits/357339dc3b6b5539d4ff698c5131caf737e7ee78, might seem like an overkill but I've also seen a case where `serverstart()` failed with a different error (#1226)  so maybe we should keep this test?

folke commented:
yep, makes sense!

ibhagwan commented:
If you feel this is ready to be merged I can ship it :-)

folke commented:
Go for it and thank you! :)

ibhagwan commented:
Tysm @folke, this has been a great help and for sure will prove very valuable in future troubleshooting .
